lib/internal/streams/writable.js

```diff
@@ -629,31 +635,28 @@ Writable.prototype.end = function(chunk, encoding, cb) {
   }
 
   if (err) {
-    // Do nothing...
-  } else if (!state.errored && !state.ending) {
+    process.nextTick(cb, err);
+  } else if (state.finished) {
+    process.nextTick(cb, new ERR_STREAM_ALREADY_FINISHED('end'));
+  } else if (state.destroyed) {
+    process.nextTick(cb, new ERR_STREAM_DESTROYED('end'));
+  } else if (state.ending) {
+    if (cb !== nop) {
+      state[kOnFinished].push(cb);
+    }
+  } else {
     // This is forgiving in terms of unnecessary calls to end() and can hide
     // logic errors. However, usually such errors are harmless and causing a
     // hard error can be disproportionately destructive. It is not always
     // trivial for the user to determine whether end() needs to be called
     // or not.
 
     state.ending = true;
-    finishMaybe(this, state, true);
-    state.ended = true;
-  } else if (state.finished) {
-    err = new ERR_STREAM_ALREADY_FINISHED('end');
-  } else if (state.destroyed) {
-    err = new ERR_STREAM_DESTROYED('end');
-  }
-
-  if (typeof cb === 'function') {
-    if (err) {
-      process.nextTick(cb, err);
-    } else if (state.finished) {
-      process.nextTick(cb, null);
-    } else {
+    if (cb !== nop) {
       state[kOnFinished].push(cb);
     }
+    finishMaybe(this, state, true);
+    state.ended = true;
   }
```

Comment:
This whole `end` section needed some restructuring, it was kind of wrong and only worked due to the unnecessary next tick